README.rst
ConfToolAPI
=======================

.. image:: https://img.shields.io/pypi/pyversions/ConfToolAPI
   :target: https://pypi.org/project/ConfToolAPI/

.. image:: https://img.shields.io/pypi/v/ConfToolAPI
   :target: https://pypi.org/project/ConfToolAPI/

.. image:: https://img.shields.io/badge/code%20style-black-000000.svg
   :target: https://github.com/psf/black
   :alt: Code style: Black

.. image:: https://img.shields.io/github/last-commit/llamm-de/conftool-python
   :target: https://github.com/llamm-de/conftool-python

.. image:: https://img.shields.io/pypi/l/ConfToolAPI
   :target: https://pypi.org/project/ConfToolAPI/

This package provides you with convenient methods to call the 
`ConfTool API <https://www.conftool.net/ctforum/index.php/topic,280.0.html>`_
and easily retrieve data from it.

First steps
------------
After installing the ConfToolAPI package,
::

   python3 -m pip install ConfToolAPI

you can retrieve data from the official ConfTool API, e.g. 
::

   from ConfToolAPI import APIHandler
    
   api = APIHandler("your_endpoint_name", "your_api_key")
   users = api.get_users()

This will deliver you the data of all users registered in ConfTool for 
conference.

If you want to get detailed information about a particular user, you 
can use the following call 
::

   user_data = api.get_user_details("username_or_email")

To check for the validity of a users credentials, you can use 
::

   result = api.check_login("username_or_email", "password")

If you want to access other data from ConfTool, please use the
``admin_export(datafield, include_deleted=False, custom_query=None)`` function 
of the ``APIHandler`` class. The datafields you may query can be one of the 
following:

* papers
* invitations
* authors
* subsumed_authors
* topics
* reviews
* reviewers
* sessions
* participants
* nametags
* identities
* event_participants
* payments
* identities 

To refine the query, a custom extension of the query can be added using the ``custom_query``
argument. For further details on the structure of such queries, please see the 
documentation of `ConfTool <https://www.conftool.net/ctforum/index.php/topic,280.0.html>`_.

License
-------
This project is licensed under the MIT license. For more details, 
please see `LICENSE.rst <LICENSE.rst>`_ file.

CHANGELOG.rst
Changelog
=========

Versions are following `SemVer <https://semver.org/lang/de/>`_ 2.0.0. 

Version: 1.0.0 (2022-06-10)
---------------------------
* Initial release

tests/test_ConfToolAPI.py
import pytest
import time

from ConfToolAPI.APIHandler import APIHandler

API_KEY_DUMMY = "aOig7Nm3B"
INITIAL_NONCE_DUMMY = 1
ENDPOINT_DUMMY = "test2022"

@pytest.fixture(scope="session")
def env_file_nonce(tmp_path_factory):
    f = tmp_path_factory.mktemp("data") / ".env_nonce"
    f.write_text(f"ENDPOINT_NAME = {ENDPOINT_DUMMY} \n API_KEY = {API_KEY_DUMMY} \n INITIAL_NONCE = {INITIAL_NONCE_DUMMY}")
    return f

@pytest.fixture(scope="session")
def env_file(tmp_path_factory):
    f = tmp_path_factory.mktemp("data") / ".env"
    f.write_text(f"ENDPOINT_NAME = {ENDPOINT_DUMMY} \n API_KEY = {API_KEY_DUMMY}")
    return f

def test_initialize_nonce():
    api = APIHandler(ENDPOINT_DUMMY, API_KEY_DUMMY)
    api_initial_nonce = APIHandler(ENDPOINT_DUMMY, API_KEY_DUMMY, initial_nonce=INITIAL_NONCE_DUMMY)

    assert api.last_nonce <= int(time.time() * 10000)
    assert api_initial_nonce.last_nonce == INITIAL_NONCE_DUMMY

def test_object_from_env_file_nonce(env_file_nonce):
    api = APIHandler.from_dotenv(env_file_nonce)
    assert api.api_key == API_KEY_DUMMY
    assert api.last_nonce == INITIAL_NONCE_DUMMY
    assert api.base_url == f"https://www.conftool.org/{ENDPOINT_DUMMY}/rest.php"

def test_object_from_env_file(env_file):
    api = APIHandler.from_dotenv(env_file)
    assert api.api_key == API_KEY_DUMMY
    assert api.last_nonce <= int(time.time() * 10000)
    assert api.base_url == f"https://www.conftool.org/{ENDPOINT_DUMMY}/rest.php"

# def test_api_call_except_for_non_xml_response():
#     assert False

# def test_api_call_except_for_error_response():
#     assert False

def test_passhash():
    api = APIHandler(ENDPOINT_DUMMY, API_KEY_DUMMY, initial_nonce=INITIAL_NONCE_DUMMY)    
    hash = api._create_passhash()

    assert hash == '775d4968b4aca5a8a9ac2f3d55393f05fb4968cb79424bb199d26e23c45c2df6'

def test_repr():
    api = APIHandler(ENDPOINT_DUMMY, API_KEY_DUMMY) 
    assert repr(api) == f"<ConfToolAPI object: https://www.conftool.org/{ENDPOINT_DUMMY}/rest.php>"

conftool/ConfToolAPI.py
import hashlib
import time
import requests
import xmltodict

class ConfToolAPI:
    """
    A simple class for handling API requests to conftool.org API.
    """

    def __init__(self, endpoint_name, api_key, initial_nonce=None) -> None:
        """
        Constructor for APIHandler
        """
        self.api_key = api_key
        self.base_url = f"https://www.conftool.org/{endpoint_name}/rest.php"
        
        if initial_nonce:
            self.last_nonce = initial_nonce
        else:
            self.last_nonce = int(time.time() * 10000)

    def _create_passhash(self):
        """
        Create API passhash according to the documentation at https://www.conftool.net/ctforum/index.php/topic,280.0.html
        """
        self.last_nonce += 1
        to_hash = f"{self.last_nonce}{self.api_key}"
        passhash = hashlib.sha256(to_hash.encode("utf-8")).hexdigest()

        return passhash

    def admin_export(self, datafield, include_deleted=False, custom_query=None):
        """
        Retrieve data from adminExport endpoint of ConfTool API
        """

        # Assemble queries in url
        if include_deleted:
            deleted_flag = 1
        else:
            deleted_flag = 0
        
        url = (f"{self.base_url}?page=adminExport&export_select={datafield}"
               f"&form_include_deleted={deleted_flag}&form_export_format=xml"
               f"&form_export_header=default&cmd_create_export=true")

        if custom_query:
            url += custom_query

        # call api
        data = self._call_api(url)

        return data

    def get_users(self, include_deleted=False, custom_query=None):
        """
        Convenience wrapper for get list of users
        """
        return self.admin_export("users", include_deleted, custom_query)

    def get_user_details(self, identifier):
        """
        Get data of a single user identified by username or email
        """
        url = (f"{self.base_url}?page=remoteLogin&user={identifier}&command=request")

        # call api
        data = self._call_api(url)

        return data

    def check_login(self, identifier, password):
        """
        Check if a user identified by username or mail can log in with given password
        """
        url = (f"{self.base_url}?page=remoteLogin&user={identifier}"
               f"&command=login&password={password}")

        # call api
        data = self._call_api(url)

        return data

    def _call_api(self, url):
        """
        Call the API, check response and return data
        """

        passhash = self._create_passhash()
        url += f"&nonce={self.last_nonce}&passhash={passhash}"
        
        response = requests.get(url)

        # Convert xml response into OrderedDictionary
        try:
            data = xmltodict.parse(response.text)
        except:
            raise RuntimeError(f'File Download is not yet implemented for ConfTool API.')

        # Check response status. 
        # Since the ConfTool API always returns status code 200 within the header, 
        # we have to manually check the error message.
        if "rest" in data:
            raise RuntimeError(f'Fetching API did not work: {data["rest"]["message"]}')

        return data

    def __repr__(self) -> str:
       return f"<ConfToolAPI object: {self.base_url}>"

ConfToolAPI/APIHandler.py
import hashlib
import time
import requests
import xmltodict
from dotenv import dotenv_values

class APIHandler:
    """
    A simple class for handling API requests to conftool.org API.
    """

    def __init__(self, endpoint_name: str, api_key: str, initial_nonce: int = None) -> None:
        """
        Constructor for APIHandler
        """
        self.api_key = api_key
        self.base_url = f"https://www.conftool.org/{endpoint_name}/rest.php"
        
        if initial_nonce:
            self.last_nonce = initial_nonce
        else:
            self.last_nonce = int(time.time() * 10000)

    @classmethod
    def from_dotenv(cls, env_file_name: str) -> 'APIHandler':
        """
        Overloading of constructor for using .env files to create new object
        """
        config = dotenv_values(env_file_name)
        if "INITIAL_NONCE" in config:
            return cls(config["ENDPOINT_NAME"], config["API_KEY"], initial_nonce=int(config["INITIAL_NONCE"]))
        else: 
            return cls(config["ENDPOINT_NAME"], config["API_KEY"])

    def _create_passhash(self) -> str:
        """
        Create API passhash according to the documentation at https://www.conftool.net/ctforum/index.php/topic,280.0.html
        """
        self.last_nonce += 1
        to_hash = f"{self.last_nonce}{self.api_key}"
        passhash = hashlib.sha256(to_hash.encode("utf-8")).hexdigest()

        return passhash

    def admin_export(self, datafield: str, include_deleted: bool = False, custom_query: str =None) -> dict:
        """
        Retrieve data from adminExport endpoint of ConfTool API
        """

        # Assemble queries in url
        if include_deleted:
            deleted_flag = 1
        else:
            deleted_flag = 0
        
        url = (f"{self.base_url}?page=adminExport&export_select={datafield}"
               f"&form_include_deleted={deleted_flag}&form_export_format=xml"
               f"&form_export_header=default&cmd_create_export=true")

        if custom_query:
            url += custom_query

        # call api
        data = self._call_api(url)

        return data

    def get_users(self, include_deleted: bool = False, custom_query: str = None) -> dict:
        """
        Convenience wrapper for get list of users
        """
        return self.admin_export("users", include_deleted, custom_query)

    def get_user_details(self, identifier: str) -> dict:
        """
        Get data of a single user identified by username or email
        """
        url = (f"{self.base_url}?page=remoteLogin&user={identifier}&command=request")

        # call api
        data = self._call_api(url)

        return data

    def check_login(self, identifier: str, password: str) -> dict:
        """
        Check if a user identified by username or mail can log in with given password
        """
        url = (f"{self.base_url}?page=remoteLogin&user={identifier}"
               f"&command=login&password={password}")

        # call api
        data = self._call_api(url)

        return data

    def _call_api(self, url: str) -> dict:
        """
        Call the API, check response and return data
        """

        passhash = self._create_passhash()
        url += f"&nonce={self.last_nonce}&passhash={passhash}"
        
        response = requests.get(url)

        # Convert xml response into OrderedDictionary
        try:
            data = xmltodict.parse(response.text)
        except:
            raise RuntimeError(f'File Download is not yet implemented for ConfTool API.')

        # Check response status. 
        # Since the ConfTool API always returns status code 200 within the header, 
        # we have to manually check the error message.
        if "rest" in data:
            raise RuntimeError(f'Fetching API did not work: {data["rest"]["message"]}')

        return data

    def __repr__(self) -> str:
       return f"<ConfToolAPI object: {self.base_url}>"